implementations/c/ockam/channel/channel_impl.c

```diff
@@ -0,0 +1,297 @@
+    goto exit;
+  } //!!onward route
+  if (0 != *p_encoded++) {
+    p_encoded = NULL;
+    goto exit;
+  } //!!return route
+exit:
+  return p_encoded;
+}
+
+ockam_error_t channel_decrypt(ockam_channel_t* p_ch,
+                              uint8_t*         p_cipher_text,
+                              size_t           cipher_text_length,
+                              uint8_t*         p_encoded_text,
+                              size_t           encoded_text_size,
+                              size_t*          p_encoded_text_length)
+{
+  ockam_error_t error = OCKAM_ERROR_NONE;
+
+  if (CHANNEL_STATE_SECURE == p_ch->state) {
+    error = ockam_key_decrypt(
+      &p_ch->key, p_encoded_text, encoded_text_size, p_cipher_text, cipher_text_length, p_encoded_text_length);
+    if (error) goto exit;
+  } else {
+    ockam_memory_copy(gp_ockam_channel_memory, p_encoded_text, p_cipher_text, cipher_text_length);
+    *p_encoded_text_length = cipher_text_length;
+  }
+exit:
+  if (error) log_error(error, __func__);
+  return error;
+}
+
+ockam_error_t channel_process_message(uint8_t* p_encoded,
+                                      size_t   encoded_text_length,
+                                      uint8_t* p_clear_text,
+                                      size_t*  p_clear_text_length)
+{
+  ockam_error_t        error        = OCKAM_ERROR_NONE;
+  codec_message_type_t message_type = *p_encoded++;
+  switch (message_type) {
+  case PING:
+    break;
+  case PAYLOAD:
+    *p_clear_text_length = encoded_text_length - sizeof(uint8_t);
+    ockam_memory_copy(gp_ockam_channel_memory, p_clear_text, p_encoded, *p_clear_text_length);
+    break;
+  default:
+    error = CHANNEL_ERROR_NOT_IMPLEMENTED;
+    goto exit;
+  }
+exit:
+  if (error) log_error(error, __func__);
+  return error;
+}
+
+ockam_error_t ockam_channel_init(ockam_channel_t* p_ch, ockam_channel_attributes_t* p_attrs)
+{
+  ockam_error_t error = OCKAM_ERROR_NONE;
+
+  if ((NULL == p_ch) || (NULL == p_attrs) || (NULL == p_attrs->reader) || (NULL == p_attrs->writer) ||
+      (NULL == p_attrs->memory)) {
+    error = CHANNEL_ERROR_PARAMS;
+    goto exit;
+  }
+
+  gp_ockam_channel_memory = p_attrs->memory;
+  p_ch->vault             = p_attrs->vault;
+
+  error = ockam_memory_alloc_zeroed(gp_ockam_channel_memory, (void**) &p_ch->channel_reader, sizeof(ockam_reader_t));
+  if (error) goto exit;
+  p_ch->channel_reader->read = channel_read;
+  p_ch->channel_reader->ctx  = p_ch;
+
+  error = ockam_memory_alloc_zeroed(gp_ockam_channel_memory, (void**) &p_ch->channel_writer, sizeof(ockam_writer_t));
+  if (error) goto exit;
+  p_ch->channel_writer->write = channel_write;
+  p_ch->channel_writer->ctx   = p_ch;
+
+  p_ch->transport_reader = p_attrs->reader;
+  p_ch->transport_writer = p_attrs->writer;
+
+  error = ockam_xx_key_initialize(
+    &p_ch->key, gp_ockam_channel_memory, p_ch->vault, p_ch->channel_reader, p_ch->channel_writer);
+
+  p_ch->state = CHANNEL_STATE_M1;
+
+exit:
```

Comment:
This could be achieved by (IMHO) much cleaner way by having several exit labels.